lib/init.php

```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Recursively find a block by attribute.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
+	 *
+	 * @return bool    True if block is found, false otherwise
+	 */
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
+		$at = 0;
+
+		/**
+		 * This is the same regex as the one used in the block parser.
+		 * It is better to use this solution to look for a block's existence
+		 * in a document compared to having to parsing the blocks in the
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
+		 */
+		$pattern = sprintf(
+			'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
+			// @TODO: we could ensure that there's a namespace on the block name.
+			preg_quote( $block_name, '~' )
+		);
+
+		while ( false !== preg_match( $pattern, $html, $matches, PREG_OFFSET_CAPTURE, $at ) ) {
+
+			// bail if no matches
+			if ( empty ($matches) ) {
+				return false;
+			}
```

Comment:
did I recommend this code? I may have, but I think `$matches` can never be empty here can it? if that _were_ the case then I believe that `preg_match()` would have returned `false` and we would have exited the loop.

if we're inside the loop it means we matched and we should be guaranteed to have an array of at least one two-element array value (the full match and the full match's offset).

Reply:
> if we're inside the loop it means we matched 

In this code, if we're inside the loop it means the matcher did not error out.

[preg_match() ](https://www.php.net/manual/en/function.preg-match.php) returns 1 if the pattern matches given subject, 0 if it does not, or false on failure. This code says:

- if the matching did not result in an error
- and it found something
- do this work

In `next_token` in WP_Block_Parse in core the checks are done more explicitly.

Do you find a problem with this check for empty matches as is?